src/solve.rs

```diff
@@ -159,56 +160,20 @@ pub async fn solve(
                 return Ok(SettledBatchAuctionModel::default());
             }
         }
+        let (approval_interaction_data, swap_interaction_data) =
+            build_payload_for_swap_and_approval(
+                &swap,
+                &query,
+                &mut tokens,
+                &mut allowances,
+                &tradable_buffer_token_list,
+                &web3,
+            )?;
 
-        let available_buffer = tokens
-            .clone()
-            .get(&query.buy_token)
-            .unwrap_or(&TokenInfoModel::default())
-            .internal_buffer
-            .unwrap_or_else(U256::zero);
-        if swap.buy_amount < available_buffer
-            && swap_tokens_are_tradable_buffer_tokens(&query, &tradable_buffer_token_list)
-        {
-            // trade only against internal buffer
-            if let Some(mut token_info) = tokens.get_mut(&query.buy_token) {
-                token_info.internal_buffer = available_buffer.checked_sub(swap.buy_amount);
-            }
-            if let Some(mut token_info) = tokens.get_mut(&query.sell_token) {
-                if let Some(buffer) = token_info.internal_buffer {
-                    token_info.internal_buffer = buffer.checked_add(swap.sell_amount);
-                } else {
-                    token_info.internal_buffer = Some(swap.sell_amount);
-                }
-            }
-        } else {
-            // use external trade
-            let spender = swap.allowance_target;
-            // Push allowance interaction data, if necessary
-            let allowance = allowances
-                .entry((query.sell_token, spender))
-                .or_insert_with(U256::zero);
-            if allowance.lt(&&mut swap.sell_amount) {
-                let token = ERC20::at(&web3, query.sell_token);
-                let method = token.approve(spender, swap.sell_amount);
-                let calldata = method.tx.data.expect("no calldata").0;
-                let interaction_item = InteractionData {
-                    target: query.sell_token,
-                    value: 0.into(),
-                    call_data: ethcontract::Bytes(calldata),
-                };
-                solution.interaction_data.push(interaction_item);
-            } else {
-                *allowance = allowance.checked_sub(swap.sell_amount).unwrap()
-            }
-            // put swap tx data into settled_batch_auction
-            let interaction_item = InteractionData {
-                target: swap.to,
-                value: swap.value,
-                call_data: ethcontract::Bytes(swap.data.0),
-            };
-            solution.interaction_data.push(interaction_item);
+        if let Some(approval_interaction_data) = approval_interaction_data {
+            solution.interaction_data.push(approval_interaction_data);
         }
-
+        solution.interaction_data.push(swap_interaction_data);
```

Comment:
First we collect all pairs `(approval, interaction)` for each trade. Then we build the solution as `(approval1, approval2, ... approvaln, interaction1, ..., interactionn)`.
However, there could be approvals for the same pair `(token, spender)`, which would mean that only the last approval will take effect and the solution is likely to revert. We could aggregate approvals in this case, or we could also just execute approvals immediately before each interactions.

Reply:
Yes, you are right.

Other solvers are also setting infinite approvals, and then the second approval would also definitively not be needed. I tried to not change this logic, as the purpose of this PR is not improving the logic, but just include internal buffer trades

Reply:
All good, for some reason I thought that allowances were all moved before the interactions, which is definitely not the case. Then all looks good, and I agree that this would just be an improvement proposal!